support $SHARNESS_TEST_SRCDIR/sharness.d directory for per-project extensions

grondo opened the pull request:
In our project that uses sharness, we want to extend sharness with project-specific test functions, modify `PATH` and other environments, and do other per-test-file setup. However, to make it easier
to stay in sync with upstream sharness, we'd like to avoid modifying sharness.sh directly.

This PR is an attempt to address this need by optionally loading "test library extensions" from
`$SHARNESS_TEST_SRCDIR/sharness.d/*.sh` if the directory exists. We support multiple
extension files in case some sharness users in the future create reusable libraries these can be
shared as is amongst interested parties by simply dropping them into the `sharness.d` directory.

I would not consider this something to submit to upstream git, as their `test-lib.sh` is only meant for
a single project. However, since the purpose of sharness is to be reused among many projects, I
thought this enhancement might be acceptable.

I also took a stab at updating `README.md`

chriscool commented:
> In our project that uses sharness, we want to extend sharness with project-specific test functions,
> modify PATH and other environments, and do other per-test-file setup. However, to make it easier
> to stay in sync with upstream sharness, we'd like to avoid modifying sharness.sh directly.

You can do that by having your own test library that sources "sharness.sh" and sourcing your own test library, instead of directly sourcing sharness.sh, in your test scripts.

For example the go-ipfs project does that, see:

https://github.com/ipfs/go-ipfs/blob/master/test/sharness/lib/test-lib.sh

grondo commented:
Yes, that works for a single test library, but I was thinking of building up a set of shareable extensions for all my projects, perhaps one per file. I agree that for single projects, your approach works and is more tractable...

chriscool commented:
Ok. Except for some indentation problems, this looks good to me. Thanks!

grondo commented:
Ok, I've rebased on current master d5e2b12b5fb8b29559159a57a72fc5bbafea670e and (hopefully) fixed up all strange indenting.

grondo commented:
Ok, another attempt at indentation fixups. Sorry if I still haven't got it quite right. My eyes are blurring now... :-)

chriscool commented:
Also could you rebase to avoid the conflicts. Thanks!

grondo commented:
Ok, rebased to current master 82c3ec8cb4c941581d89637b74488a1d5a2f40a4 and removed the accidental debugging `set -x` (good catch)

Thanks for your patience.

chriscool commented:
It looks like if the 'sharness.d' directory exists but there is no '*.sh' file in it, the test scripts will error out immediately.

grondo commented:
Oh, bummer, sorry about that! Nice catch.
Would it be acceptable to replace that glob with an invocation of `find(1)`?
Otherwise, if you have another tractable solution I'd be willing to do whatever.

Also, seems like a test for this case should be added

grondo commented:
Ok, @chriscool, I've taken the approach suggested by [Wheeler](http://www.dwheeler.com/essays/filenames-in-shell.html) and simply check for existence of `"$file"` before sourcing as a portable way to handle empty glob match without use of bash extensions or `find(1)`

I also added a test for the failing case.

Let me know if this is acceptable and I could squash these two new commits into the originals. I just thought it would be easier to check out the changes by keeping them separate for now.

Thanks!

chriscool commented:
Except for the `test` versus `[ ... ]` issue above, this looks good and you can squash those two new commits into the originals. Thanks!

grondo commented:
Great. Sorry about the `test` vs `[...]` thing.
I've fixed that and squashed the test commits and sharness.sh commits together.

grondo commented:
Ok, updated to break out of loop before issuing debug message

chriscool commented:
Merged, thanks!

grondo commented:
Great, thanks @chriscool!
